reconcile/cna/integration.py

```diff
@@ -0,0 +1,133 @@
+from collections import defaultdict
+from typing import Iterable, Mapping, Optional
+from reconcile.cna.assets.null import NullAsset
+from reconcile.cna.client import CNAClient
+from reconcile.cna.state import State
+
+from reconcile.utils import gql
+from reconcile.utils.ocm_base_client import OCMBaseClient
+from reconcile.gql_definitions.cna.queries.cna_provisioners import (
+    CNAProvisionerV1,
+    query as cna_provisioners_query,
+)
+from reconcile.gql_definitions.cna.queries.cna_resources import (
+    CNANullAssetV1,
+    NamespaceCNAssetV1,
+    NamespaceV1,
+    query as namespaces_query,
+)
+from reconcile.typed_queries.app_interface_vault_settings import (
+    get_app_interface_vault_settings,
+)
+from reconcile.utils.secret_reader import SecretReaderBase, create_secret_reader
+from reconcile.utils.semver_helper import make_semver
+
+
+QONTRACT_INTEGRATION = "cna_resources"
+QONTRACT_INTEGRATION_VERSION = make_semver(0, 1, 0)
+QONTRACT_TF_PREFIX = "qrtfcf"
+
+
+class CNAConfigException(Exception):
+    pass
+
+
+class CNAIntegration:
+    def __init__(
+        self,
+        cna_clients: Mapping[str, CNAClient],
+        namespaces: Iterable[NamespaceV1],
+        desired_states: Optional[Mapping[str, State]] = None,
+        actual_states: Optional[Mapping[str, State]] = None,
+    ):
+        self._cna_clients = cna_clients
+        self._namespaces = namespaces
+        self._desired_states = desired_states if desired_states else defaultdict(State)
+        self._actual_states = actual_states if actual_states else defaultdict(State)
+
+    def assemble_desired_states(self):
+        self._desired_states = defaultdict(State)
+        for namespace in self._namespaces:
+            for provider in namespace.external_resources or []:
+                # TODO: this should probably be filtered within the query already
+                if not isinstance(provider, NamespaceCNAssetV1):
+                    continue
+                for resource in provider.resources or []:
+                    if isinstance(resource, CNANullAssetV1):
```

Comment:
we discussed this briefly. having a mapper between the types representing CNA modules and the app-interfaces types would enable us to make this generic. but since it is just this one class here, we can do that in a followup PR

Reply:
Fully agree. We can enhance this when we add a second kind of asset.